### ISSUE TITLE
ENH: Faster inverse-Wishart sampling

### ISSUE THREAD

**Issue by sethaxen** [scipy/scipy#19432]
### Is your feature request related to a problem? Please describe.

Currently inverse-Wishart sampling ([here](https://github.com/scipy/scipy/blob/d7f12d38b171f3a6ae841937bb6c2444a1e04b8d/scipy/stats/_multivariate.py#L2957-L3004)) Cholesky decomposes the scale matrix to invert it, then Cholesky decomposes its inverse, samples a Cholesky factor from the Wishart distribution, then inverts it to form an inverse-Wishart matrix.

Algorithm 5 in https://arxiv.org/abs/2310.15884v1 generates a Cholesky factor of an inverse-Wishart matrix directly. Table 1 shows that for the $\Psi$ parameterization used in `invwishart`, this algorithm eliminates 1 triangular product, 1 triangular solve, and 1 Cholesky factorization compared to the current algorithm, while adding no additional operations.

### Describe the solution you'd like.

I propose adopting the new algorithm.

From a cursory glance, it seems `invwishart_gen` might subclass `wishart_gen` primarily so they can share the same `_standard_rvs` method. The new algorithm would have a different `_standard_rvs` method, so I wonder if it would make sense for `invwishart_gen` to subtype `wishart_gen` anymore.

### Additional context (e.g. screenshots, GIFs)

Here are the relevant sections of the paper:
<img src=https://github.com/scipy/scipy/assets/8673634/df12cd40-e673-455c-beee-5934393394b7 width=400>

<img src=https://github.com/scipy/scipy/assets/8673634/daab97bc-abc4-4852-a6f3-71e4617e1d52 width=400>

And here is a reference implementation: https://github.com/sethaxen/rand_invwishart/tree/main/RandomInverseWishart.jl

**Comment by ilayn** [comment]
Don't know what this means on the statistics side but if you use `trsm` in the step 8 of alg 4. it would be more accurate and faster.

**Comment by sethaxen** [comment]
Step 7 uses `trtri`, while Step 8 uses `trmm`. I chose this over combining to `trsm` because in benchmarks it was slightly faster. It also allows for the entire algorithm to be in-place; `trsm` would require allocating an additional matrix for storage.

**Comment by ilayn** [comment]
That should not be the case [dtrsm](https://www.netlib.org/lapack/explore-html/d1/d54/group__double__blas__level3_ga6a0a7704f4a747562c1bd9487e89795c.html) combines both of those steps without explicitly forming the multiplication and does it in-place overwriting $U_\Omega$. No allocation should be necessary.

**Comment by sethaxen** [comment]
Yes, but in Algorithm 4, $U_\Omega$ is a user-provided parameter and should probably not be overwritten. However, in Algorithm 5, if $\Omega$ is provided, then $U_\Omega$ once computed could be overwritten. I just benchmarked, and using `trsm` in that case is indeed slightly faster than `trtri` and `trmm`. Probably not relevant for scipy, since either way it would make sense to use `trsm` here, but relevant for the paper.

I don't have a great sense for how relevant the improved accuracy due to `trsm` would be when `Z` is random.

**Comment by dschmitz89** [comment]
Could you benchmark your algorithm against what is currently implemented in scipy? Reducing the number of linear algebra operations sounds in principle good although I do not know the LAPACK functions.

**Comment by sethaxen** [comment]
Sure, I implemented the algorithm in https://github.com/sethaxen/scipy/blob/invwishart_rvs_direct/scipy/stats/_multivariate.py#L2957-L3087 and ran the below benchmark on main and this branch:

```python
import timeit
import numpy as np
from scipy.stats import invwishart

# Seed for reproducibility
random_seed = 42
np.random.seed(random_seed)

# Define the parameters for the inverse Wishart distribution
dim = 1_000  # dim
df = dim     # degrees of freedom

def rand_pd_mat(dim):
    _, R = np.linalg.qr(np.random.normal(size=(dim, dim)))
    return R.T @ R

scale = rand_pd_mat(dim)
n_samples = 10

def generate_samples():
    invwishart.rvs(df=df, scale=scale, size=n_samples)

# Time the function
number_of_evals = 1
total_time = timeit.repeat(generate_samples, repeat=100, number=number_of_evals)
min_time_per_run = min(total_time) / number_of_evals
print(f"min time per run: {min_time_per_run}")
# old: 0.6621020569873508
# new: 0.39741522597614676
```
So at this size the new algorithm takes ~60% the runtime as the current one.

**Comment by dschmitz89** [comment]
> Sure, I implemented the algorithm in https://github.com/sethaxen/scipy/blob/invwishart_rvs_direct/scipy/stats/_multivariate.py#L2957-L3087 and ran the below benchmark on main and this branch:
> 
> ```python
> import timeit
> import numpy as np
> from scipy.stats import invwishart
> 
> # Seed for reproducibility
> random_seed = 42
> np.random.seed(random_seed)
> 
> # Define the parameters for the inverse Wishart distribution
> dim = 1_000  # dim
> df = dim     # degrees of freedom
> 
> def rand_pd_mat(dim):
>     _, R = np.linalg.qr(np.random.normal(size=(dim, dim)))
>     return R.T @ R
> 
> scale = rand_pd_mat(dim)
> n_samples = 10
> 
> def generate_samples():
>     invwishart.rvs(df=df, scale=scale, size=n_samples)
> 
> # Time the function
> number_of_evals = 1
> total_time = timeit.repeat(generate_samples, repeat=100, number=number_of_evals)
> min_time_per_run = min(total_time) / number_of_evals
> print(f"min time per run: {min_time_per_run}")
> # old: 0.6621020569873508
> # new: 0.39741522597614676
> ```
> 
> So at this size the new algorithm takes ~60% the runtime as the current one.

Nice :). PR welcome then, I would say. About the linear algebra operations: we try to avoid explicit inversions, and it looks like your code does not perform any, so this should be good. For the review of the LAPACK related things, could you ping @ilayn on the PR (if @ilayn  does not have too much on his plate currently)? Most stats maintainers are not super familiar with LAPACK.

**Comment by sethaxen** [comment]
Sure, happy to open a PR!

**Comment by mdhaber** [comment]
Could we also identify a maintainer to review the rest, or perhaps @ilayn is willing to review and merge?

@sethaxen has your paper been peer reviewed, or is there another source of the algorithm?

**Comment by ilayn** [comment]
Yes I can take a look. Doesn't look stats heavy (fingers crossed 😃)

**Comment by sethaxen** [comment]
> @sethaxen has your paper been peer reviewed, or is there another source of the algorithm? 

No it hasn't; this preprint introduced the algorithm.

**Comment by sethaxen** [comment]
> or is there another way we can confirm that samples will be consistent with the distribution (besides numerically)?

Someone would need to check the math. In particular the theorem that derived the change of variables for the triangular inverse. Although that can also be checked numerically by checking the Jacobian determinants of the change of variables. I've already done this.

**Comment by sethaxen** [comment]
Here's a numerical check that the log-density of $Z$ implied by Theorem 3.2 agrees with the change of variables we get by explicitly transforming the inverse-Wishart to $Z$ and computing the log-det-Jacobian using autodiff:

```python
import numpy as np
from scipy.stats import chi, norm, invwishart
import jax

def triu_to_vec(x):
    y = x[np.triu_indices_from(x)]
    return y

def vec_to_triu_jax(y):
    dim = (int((8 * len(y) + 1)**0.5) - 1) // 2
    x = jax.numpy.zeros((dim, dim), dtype=jax.numpy.float64)
    x = x.at[np.triu_indices(dim)].set(y)
    return x

def transform(Z, UOmega):
    UB = jax.scipy.linalg.solve_triangular(Z, UOmega)
    B = UB.T @ UB
    return B

def rand_chol_upper(dim):
    Z = np.linalg.qr(np.random.randn(dim, dim))[1]
    Z[np.diag_indices(dim)] = np.fabs(Z[np.diag_indices(dim)])
    return Z

def logpdf_direct(Z, df):
    dim = Z.shape[0]
    lp = norm.logpdf(Z[np.triu_indices(dim, 1)]).sum()
    lp += chi.logpdf(np.diag(Z), df=df - dim + np.arange(dim) + 1).sum()
    return lp

def logpdf_change_of_variables(Z, df, UOmega):
    B = transform(Z, UOmega)
    f = lambda z: triu_to_vec(transform(vec_to_triu_jax(z), Uscale))
    J = jax.jacfwd(f)(triu_to_vec(Z))
    logdetJ = np.linalg.slogdet(J)[1]
    Omega = UOmega.T @ UOmega
    lp = invwishart.logpdf(B, df=df, scale=Omega) + logdetJ
    return lp

for dim in range(1, 11):
    for df in (dim, dim + 3):
        Z = rand_chol_upper(dim)
        Uscale = rand_chol_upper(dim)
        lp1 = logpdf_direct(Z, df)
        lp2 = logpdf_change_of_variables(Z, df, Uscale)
        assert np.isclose(lp1, lp2)
```

**Comment by sethaxen** [comment]
I noticed `invwishart_gen._logpdf` also Cholesky factorizes `scale`,  Cholesky factorizes `x` to invert it to `xinv`, then Cholesky factorizes `xinv` again to compute the logpdf. We can at least eliminate one factorization using the transform in the paper (without the change of variables), so I'll add that to the PR as well with a benchmark.
